community/cypher/cypher-compiler-3.2/src/main/scala/org/neo4j/cypher/internal/compiler/v3_2/pipes/DistinctVarLengthExpandPipe.scala

```diff
@@ -0,0 +1,268 @@
+                                 relName: String,
+                                 toName: String,
+                                 types: LazyTypes,
+                                 dir: SemanticDirection,
+                                 min: Int,
+                                 max: Int,
+                                 filteringStep: VarLengthPredicate = VarLengthPredicate.NONE)
+                                (val estimatedCardinality: Option[Double] = None)
+                                (implicit pipeMonitor: PipeMonitor) extends PipeWithSource(source, pipeMonitor) with RonjaPipe {
+  self =>
+
+  trait State {
+    // Note that the Node part here can be null.
+    // This code is used in a hot spot, and avoiding object creation is important.
+    def next(): (State, ExecutionContext)
+  }
+
+  case object Empty extends State {
+    override def next() = (this, null)
+  }
+
+  class FullExpandDepths(val rels: Array[Relationship]) {
+    val depths = new Array[Int](rels.length)
+
+    def minExcept(relId: Long): Int = {
+      var min = Int.MaxValue
+      var i = 0
+      while (i < rels.length) {
+        if (rels(i).getId != relId)
+          min = math.min(depths(i), min)
+        i += 1
+      }
+      min
+    }
+  }
+
+  trait CheckPath {
+    def pathLength: Int
+
+    val path: Array[Long]
+
+    def seenRelationshipInPath(r: Long): Boolean = {
+      if (pathLength == 0) return false
+      var idx = 0
+      while (idx < pathLength) {
+        if (path(idx) == r) return true
+        idx += 1
+      }
+      false
+    }
+  }
+
+  class PrePruningDFS(
+             whenEmptied: State,
+             node: Node,
+             val path: Array[Long],
+             val pathLength: Int,
+             val state: QueryState,
+             row: ExecutionContext,
+             expandMap: PrimitiveLongObjectMap[FullExpandDepths]
+  ) extends State with Expandable with CheckPath {
+
+    var rels: Iterator[Relationship] = _
+
+    override def next(): (State, ExecutionContext) = {
+
+      if (pathLength < self.max) {
+        if (rels == null)
+          rels = expand(row, node)
+
+        while (rels.hasNext) {
+          val r = rels.next()
+          val relId = r.getId
+
+          if (!seenRelationshipInPath(relId)) {
+            val nextNode = r.getOtherNode(node)
+            path(pathLength) = relId
+            if (pathLength + 1 >= self.min)
+              return new PruningDFS(
+                                  this,
+                                  nextNode,
+                                  path,
+                                  pathLength + 1,
+                                  state,
+                                  row,
+                                  expandMap,
+                                  depth => {}
+                                ).next()
+            else
+              return new PrePruningDFS(this, nextNode, path, pathLength + 1, state, row, expandMap).next()
```

Comment:
Just realized this might be tail recursive. If that's intended please add an annotation.